### PR TITLE
Fix handling of top-level exceptions

### LINKED ISSUE #4357: [CT-461] Capture exceptions at top level of threads and cli
### Is there an existing feature request for this?

- [X] I have searched the existing issues

### Describe the Feature

discovered while coding up #4326. Somewhere in [here](https://github.com/dbt-labs/dbt-core/blob/a9b4316346e3c25b82fc59fa1f090e14b57a65db/core/dbt/adapters/base/impl.py#L191-L202) when an exception is raised, the thread just hangs and you have to step through a debugger to read the exception message. Exceptions should bubble all the way to the top even in threaded environments.

you can recreate this by making `fields_to_json` throw.

### LINKED ISSUE #5564: [CT-943] Captured top-level exception do not store into log or display stacktrace.
Because of the flags associated with the MainEncounteredErroor and MainStackTrace events the stack trace was neither sent to stdout nor written to a file, and neither event was written to the log.

### PR DESCRIPTION
**Pull request by gshank**
resolves #5564

### Description

Errors that were caught as BaseExceptions in the main function were not displaying or storing the stack trace because it was flagged as debug level and also flagged as not to be written to the file. I have removed the "NoFile" flag from the events that handle top level exceptions and left it only for integration tests and "clean" command events.

A different pr will be used to address catching exception in threads.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: core/dbt/main.py [hunk 1]
```diff
@@ -142,7 +142,6 @@ def main(args=None):
             exit_code = e.code
 
         except BaseException as e:
-            traceback.print_exc()
```
Comment (jtcohen6) [inline core/dbt/main.py:145]: Given that this was an incidental inclusion in v1.2.0, I'd also be in favor of a PR that removed this line alone, and backporting that PR to `1.2.latest` for inclusion in a v1.2.1 patch release

Comment (gshank) [reply core/dbt/main.py:145]: I'm actually looking at two different problems, really, the lack of a stack trace for top level exceptions and catching exceptions in threads. I could make the current code the "lack of a stack trace" fix and backport it?

Comment (jtcohen6) [reply core/dbt/main.py:145]: I like that idea!

Comment (nathaniel-may) [reply core/dbt/main.py:145]: good plan.

**Review by ChenyuLInx — APPROVED** [review]
Looks good to me!
No blocker but do we want to make this available under a flag so engineers/users can access more trace when they need to?

**Pull request merged**